### DISCUSSION TITLE
I am trying the SavedModel format export

### DISCUSSION THREAD

**Discussion by famy040326** [mindee/doctr#1086] [Q&A]
With the code below, I was able to output the db_resnet50 model in SavedModel format. Thank you.
[](https://github.com/mindee/doctr/discussions/1079)

```
import tensorflow as tf
from doctr.models import db_resnet50
model = db_resnet50(pretrained=True, exportable=True)
input_t = tf.random.uniform(shape=[1, 1024, 1024, 3], maxval=1, dtype=tf.float32)
_ = model(input_t, training=False)
tf.saved_model.save(model, '. /doctr_SavedModel/')
```

I tried to output the model of crnn_vgg16_bn in SavedModel format with the following code in the same way.
However, I get an error. How can I solve this problem?

```
import tensorflow as tf
from doctr.models import crnn_vgg16_bn
model = crnn_vgg16_bn(pretrained=True, exportable=True)
input_t = tf.random.uniform(shape=[1, 32, 128, 3], maxval=1, dtype=tf.float32)
_ = model(input_t, training=False)
tf.saved_model.save(model, '. /doctr_SavedModel/')

```

> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\saved_model\save.py", line 1334, in save
>     save_and_return_nodes(obj, export_dir, signatures, options)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\saved_model\save.py", line 1369, in save_and_return_nodes
>     _build_meta_graph(obj, signatures, options, meta_graph_def))
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\saved_model\save.py", line 1536, in _build_meta_graph
>     return _build_meta_graph_impl(obj, signatures, options, meta_graph_def)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\saved_model\save.py", line 1477, in _build_meta_graph_impl
>     signatures = signature_serialization.find_function_to_export(
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\saved_model\signature_serialization.py", line 103, in find_function_to_export
>     for name, child in children:
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\saved_model\save.py", line 133, in list_children
>     super(_AugmentedGraphView, self).list_children(
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\training\tracking\graph_view.py", line 256, in list_children
>     in obj._trackable_children(save_type, **kwargs).items()]
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\training\tracking\base.py", line 1479, in _trackable_children
>     return self._get_legacy_saved_model_children(cache)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\training\tracking\base.py", line 1490, in _get_legacy_saved_model_children
>     functions = self._list_functions_for_serialization(serialization_cache)
>   File "C:\Python_dev\env1\lib\site-packages\keras\engine\training.py", line 3081, in _list_functions_for_serialization
>     functions = super(
>   File "C:\Python_dev\env1\lib\site-packages\keras\engine\base_layer.py", line 3168, in _list_functions_for_serialization
>     return (self._trackable_saved_model_saver
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\base_serialization.py", line 93, in list_functions_for_serialization
>     fns = self.functions_to_serialize(serialization_cache)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\layer_serialization.py", line 72, in functions_to_serialize
>     return (self._get_serialized_attributes(
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\layer_serialization.py", line 88, in _get_serialized_attributes
>     object_dict, function_dict = self._get_serialized_attributes_internal(
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\model_serialization.py", line 56, in _get_serialized_attributes_internal
>     super(ModelSavedModelSaver, self)._get_serialized_attributes_internal(
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\layer_serialization.py", line 98, in _get_serialized_attributes_internal
>     functions = save_impl.wrap_layer_functions(self.obj, serialization_cache)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\save_impl.py", line 197, in wrap_layer_functions
>     fn.get_concrete_function()
>   File "C:\Users\Y.FUJIZONO-BSW\AppData\Local\Programs\Python\Python39\lib\contextlib.py", line 126, in __exit__
>     next(self.gen)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\save_impl.py", line 359, in tracing_scope
>     fn.get_concrete_function(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\eager\def_function.py", line 1264, in get_concrete_function
>     concrete = self._get_concrete_function_garbage_collected(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\eager\def_function.py", line 1255, in _get_concrete_function_garbage_collected
>     concrete = self._stateful_fn._get_concrete_function_garbage_collected(  # pylint: disable=protected-access
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\eager\function.py", line 3036, in _get_concrete_function_garbage_collected
>     graph_function, _ = self._maybe_define_function(args, kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\eager\function.py", line 3292, in _maybe_define_function
>     graph_function = self._create_graph_function(args, kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\eager\function.py", line 3130, in _create_graph_function
>     func_graph_module.func_graph_from_py_func(
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\framework\func_graph.py", line 1161, in func_graph_from_py_func
>     func_outputs = python_func(*func_args, **func_kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\eager\def_function.py", line 677, in wrapped_fn
>     out = weak_wrapped_fn().__wrapped__(*args, **kwds)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\save_impl.py", line 572, in wrapper
>     ret = method(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\utils.py", line 168, in wrap_with_training_arg
>     return control_flow_util.smart_cond(
>   File "C:\Python_dev\env1\lib\site-packages\keras\utils\control_flow_util.py", line 105, in smart_cond
>     return tf.__internal__.smart_cond.smart_cond(
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\framework\smart_cond.py", line 53, in smart_cond
>     return true_fn()
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\utils.py", line 169, in <lambda>
>     training, lambda: replace_training_and_call(True),
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\utils.py", line 166, in replace_training_and_call
>     return wrapped_call(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\save_impl.py", line 652, in call
>     return call_and_return_conditional_losses(inputs, *args, **kwargs)[0]
>     return self.wrapped_call(*args, **kwargs)
>     return self.wrapped_call(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\tensorflow\python\util\traceback_utils.py", line 153, in error_handler
>     raise e.with_traceback(filtered_tb) from None
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\save_impl.py", line 572, in wrapper
>     ret = method(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\utils.py", line 168, in wrap_with_training_arg
>     return control_flow_util.smart_cond(
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\utils.py", line 169, in <lambda>
>     training, lambda: replace_training_and_call(True),
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\utils.py", line 166, in replace_training_and_call
>     return wrapped_call(*args, **kwargs)
>   File "C:\Python_dev\env1\lib\site-packages\keras\saving\saved_model\save_impl.py", line 634, in call_and_return_conditional_losses
>     call_output = layer_call(*args, **kwargs)
>   File "C:\Python_dev\env1\doctr\doctr\models\recognition\crnn\tensorflow.py", line 195, in call
>     raise ValueError("Need to provide labels during training")
> ValueError: Need to provide labels during training

**Comment by felixdittrich92** [comment]
Hi @famy040326 :wave:,
after a short test anything while calling save sets training to True (then it raises the error in our code)

```
False <-- our call
False <-- our call
False <-- now tf interanl calls
False
True <-- raises
False
True
True
False
True
False
True
False
True
```

uncomment in our code will solve your problem (not beauty):
https://github.com/mindee/doctr/blob/3d7c8127bc8df8e28ab28981474d291446df06d4/doctr/models/recognition/crnn/tensorflow.py#L195

A bit weird with the Onnx export i haven't faiced this issue and the pipe is also keras model -> saved model -> onnx :thinking:

**Reply by famy040326** [reply]
Thanks for getting back to me.
I will try to respond with an uncomment.

My question is, can you tell me the intent of the following code as it is not clear to me?

```
if kwargs.get("training", False) and target is None:
raise ValueError("Need to provide labels during training")
```

**Reply by felixdittrich92** [reply]
Yeah we have to ensure that if the model in training mode there are targets provided while training as ground truth data to compute the loss. 
While inference we pass only "images" into the model (training is set to False) to predict "words"
